# SQL - анализ базы данных для предложения нового продукта.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

# Цель:

Проанализировать базу данных для предложения нового продукта.


# План:

1. Загрузка данных, изучение общей информации

2. Подсчёт количества книг, вышедших после 1 января 2000 года

3. Подсчёт количества обзоров и средней оценки для каждой книги

4. Определение издательства, которое выпустило наибольшее число книг толще 50 страниц - так мы исключим из анализа брошюры

5. Определение автора с самой высокой средней оценкой книг— учитывая только книги с 50 и более оценками

6. Определение среднего количества обзоров от пользователей, которые поставили больше 50 оценок

7. Общий вывод по решенным задачам и рекомендации

# Загрузка данных, изучение общей информации

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

чтобы выполнить SQL-запрос, используем Pandas
pd.io.sql.read_sql(query, con = engine)

`books` - таблица с данными о книгах

In [2]:
query = '''

SELECT *
FROM books
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

`authors` - таблица с данными об авторах

In [3]:
query = '''

SELECT *
FROM authors
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

`publishers` - таблица с данными об издательствах

In [4]:
query = '''

SELECT *
FROM publishers
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

`ratings` - таблица с данными о пользовательских оценках книг

In [5]:
query = '''

SELECT *
FROM ratings
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

`reviews` - таблица с данными о пользовательских обзорах

In [6]:
query = '''

SELECT *
FROM reviews
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:

# какие ещё таблицы доступны в базе
display(pd.io.sql.read_sql('''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

''', con = engine))


# смотрим на типы столбцов в интересующих таблицах
display(pd.io.sql.read_sql('''
SELECT 
    table_name, 
    column_name, 
    data_type, 
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS 
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
''', con = engine))

schemaname           tablename         tableowner tablespace  hasindexes  \
0      public              orders    praktikum_admin       None        True   
1      public              visits    praktikum_admin       None        True   
2      public  advertisment_costs    praktikum_admin       None        True   
3      public             authors    praktikum_admin       None        True   
4      public              second  praktikum_student       None       False   
5      public            second_b  praktikum_student       None       False   
6      public          publishers    praktikum_admin       None        True   
7      public              author  praktikum_student       None        True   
8      public             reviews    praktikum_admin       None        True   
9      public             ratings    praktikum_admin       None        True   
10     public               books    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
0      False        False        False  
1      False        False        False  
2      False        False        False  
3      False         True        False  
4      False        False        False  
5      False        False        False  
6      False         True        False  
7      False        False        False  
8      False         True        False  
9      False         True        False  
10     False         True        False

table_name       column_name data_type is_nullable
0      authors         author_id   integer          NO
1      authors            author      text         YES
2   publishers      publisher_id   integer          NO
3   publishers         publisher      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

# Подсчёт количества книг, вышедших после 1 января 2000 года

In [8]:
query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

- Всего после 1 января 2000 года вышло 819 книг.

# Подсчёт количества обзоров и средней оценки для каждой книги

In [9]:
query = '''

SELECT b.book_id, 
       b.title, 
       COUNT(DISTINCT rv.review_id) AS cnt_reviews, 
       ROUND(AVG(r.rating),3) AS avg_rating
FROM books AS b

LEFT JOIN reviews AS rv
ON b.book_id = rv.book_id

LEFT JOIN  ratings AS r
ON b.book_id = r.book_id

GROUP BY b.book_id, b.title
ORDER BY cnt_reviews DESC

'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0         3.663  
1         3.977  
2         4.207  
3         4.415  
4         4.081  
..          ...  
995       3.667  
996       3.000  
997       5.000  
998       4.000  
999       4.000  

[1000 rows x 4 columns]

Всего в каталоге 1000 книг. 
Количество обзоров от 0 до 7. 
Лидером по количеству отзывов стало произведение Twilight(Twilight #1)-avg=3.66



# Определение издательства, которое выпустило наибольшее число книг толще 50 страниц - так мы исключим из анализа брошюры

In [10]:
query = '''

SELECT publisher,
    COUNT(book_id) AS number_of_books
    
FROM publishers p

JOIN books b ON b.publisher_id = p.publisher_id

GROUP BY publisher
HAVING MIN(num_pages) > 50
ORDER BY number_of_books DESC
LIMIT 1
           
'''

pd.io.sql.read_sql(query, con = engine) 

publisher  number_of_books
0  Penguin Books               42

- Penguin Books(42 выпущенных книг) - издательство, которое выпустило наибольшее число книг толще 50 страниц

# Определение автора с самой высокой средней оценкой книг— учитывая только книги с 50 и более оценками

In [11]:
query = '''

WITH 
tab AS (
SELECT book_id
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) > 50)

SELECT a.author_id, 
       a.author, 
       AVG(r.rating) AS avg_rating
FROM books AS b

INNER JOIN tab
ON b.book_id = tab.book_id

INNER JOIN authors AS a
ON b.author_id = a.author_id

INNER JOIN ratings AS r
ON tab.book_id = r.book_id

GROUP BY a.author_id, a.author
ORDER BY AVG(r.rating) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

- Автор, с самой высокой средней оценкой книг с учетом книг с 50 и более оценками - J.K. Rowling/Mary GrandPré - avg-4.28

# Определение среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''

SELECT COUNT(DISTINCT rv.text)/ COUNT(DISTINCT rv.username) AS avg_rv_count
FROM ratings AS r

INNER JOIN reviews AS rv
ON r.username = rv.username

WHERE r.username in (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50)

'''

pd.io.sql.read_sql(query, con = engine)

avg_rv_count
0            24

- Cреднеее количество обзоров от пользователей, которые поставили больше 50 оценок - 24

# Общий вывод по решенным задачам и рекомендации

Исходя из данных, можно порекомендовать следующее:

- Обратить внимание на произведение "Twilight", так как оно является лидером по количеству отзывов. Возможно, это значит, что книга интересна и любима многими читателями.

- Посмотреть книги издательства Penguin Books, так как это издательство выпустило наибольшее число книг толще 50 страниц. Это может говорить о том, что издательство выбирает качественные произведения или имеет широкий выбор жанров и авторов.

- Обратить внимание на произведения автора J.K. Rowling совместно с иллюстратором Mary GrandPré, так как они имеют самую высокую среднюю оценку книг с учетом количества оценок. Это может говорить о том, что произведения данного автора и иллюстратора качественные и любимые многими читателями.

- Значительное среднее количество обзоров от пользователей, которые поставили больше 50 оценок (24), говорит о том, что читатели, которые вносят большой вклад в рейтинг книг, также оставляют много отзывов. 
